server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -0,0 +1,449 @@
+        clusterSettings.addSettingsUpdateConsumer(CLUSTER_ROUTING_ALLOCATION_AWARENESS_ATTRIBUTE_SETTING, this::setAwarenessAttributes);
+
+        setForcedAwarenessAttributes(CLUSTER_ROUTING_ALLOCATION_AWARENESS_FORCE_GROUP_SETTING.get(settings));
+        clusterSettings.addSettingsUpdateConsumer(
+            CLUSTER_ROUTING_ALLOCATION_AWARENESS_FORCE_GROUP_SETTING,
+            this::setForcedAwarenessAttributes
+        );
+    }
+
+    private void setAwarenessAttributes(List<String> awarenessAttributes) {
+        this.awarenessAttributes = awarenessAttributes;
+    }
+
+    private void setForcedAwarenessAttributes(Settings forceSettings) {
+        Map<String, List<String>> forcedAwarenessAttributes = new HashMap<>();
+        Map<String, Settings> forceGroups = forceSettings.getAsGroups();
+        for (Map.Entry<String, Settings> entry : forceGroups.entrySet()) {
+            List<String> aValues = entry.getValue().getAsList("values");
+            if (aValues.size() > 0) {
+                forcedAwarenessAttributes.put(entry.getKey(), aValues);
+            }
+        }
+        this.forcedAwarenessAttributes = forcedAwarenessAttributes;
+    }
+
+    /**
+     * Starts the new decommission request and registers the metadata with status as {@link DecommissionStatus#INIT}
+     * or the last known status if not {@link DecommissionStatus#FAILED}
+     * Once the status is updated, it tries to exclude to-be-decommissioned cluster manager nodes from Voting Configuration
+     *
+     * @param decommissionAttribute register decommission attribute in the metadata request
+     * @param listener register decommission listener
+     */
+    public void startDecommissionAction(
+        final DecommissionAttribute decommissionAttribute,
+        final ActionListener<ClusterStateUpdateResponse> listener
+    ) {
+        // register the metadata with status as DECOMMISSION_INIT as first step
+        clusterService.submitStateUpdateTask("decommission [" + decommissionAttribute + "]", new ClusterStateUpdateTask(Priority.URGENT) {
+            @Override
+            public ClusterState execute(ClusterState currentState) throws Exception {
+                // validates if correct awareness attributes and forced awareness attribute set to the cluster before starting action
+                validateAwarenessAttribute(decommissionAttribute, awarenessAttributes, forcedAwarenessAttributes);
+
+                Metadata metadata = currentState.metadata();
+                Metadata.Builder mdBuilder = Metadata.builder(metadata);
+                DecommissionAttributeMetadata decommissionAttributeMetadata = metadata.custom(DecommissionAttributeMetadata.TYPE);
+                // check if the same attribute is requested for decommission and currently not FAILED,
+                // then return the current state as is; as we don't need a state update here
+                if (isSameNonFailedRequest(decommissionAttributeMetadata, decommissionAttribute)) {
```

Comment:
nit: rename `isRetryFailed` with the negation

Reply:
Updated logic as well to improve readability -
```
private static void ensureEligibleRequest(
        DecommissionAttributeMetadata decommissionAttributeMetadata,
        DecommissionAttribute requestedDecommissionAttribute
    ) {
        String msg = null;
        if (decommissionAttributeMetadata != null) {
            // check if the same attribute is registered and handle it accordingly
            if (decommissionAttributeMetadata.decommissionAttribute().equals(requestedDecommissionAttribute)) {
                switch (decommissionAttributeMetadata.status()) {
                    case INIT:
                    case FAILED:
                        break;
                    case IN_PROGRESS:
                    case SUCCESSFUL:
                        msg = "same request is already in status [" + decommissionAttributeMetadata.status() + "], please wait for it to complete";
                        break;
                    default:
                        throw new IllegalStateException("unknown status [" + decommissionAttributeMetadata.status() + "] currently registered in metadata");
                }
            }
            else {
                switch (decommissionAttributeMetadata.status()) {
                    case SUCCESSFUL:
                        // one awareness attribute is already decommissioned. We will reject the new request
                        msg = "one awareness attribute ["
                            + decommissionAttributeMetadata.decommissionAttribute().toString()
                            + "] already successfully decommissioned, recommission before triggering another decommission";
                        break;
                    case IN_PROGRESS:
                    case INIT:
                        // it means the decommission has been initiated or is inflight. In that case, will fail new request
                        msg = "there's an inflight decommission request for attribute ["
                            + decommissionAttributeMetadata.decommissionAttribute().toString()
                            + "] is in progress, cannot process this request";
                        break;
                    case FAILED:
                        break;
                    default:
                        throw new IllegalStateException("unknown status [" + decommissionAttributeMetadata.status() + "] currently registered in metadata");
                }
            }
        }

        if (msg != null) {
            throw new DecommissioningFailedException(requestedDecommissionAttribute, msg);
        }
    }
```